### PR TITLE
JENA-967: afn:Sprintf implementation

### PR DESCRIPTION
**Pull request by ales004**
Hi,
I implemented the afn:sprintf function. In order to implement it, I had to:

- create a sprintf.java class in the sparql function part.

- implement in the XSDFuncOp.java a function dealing with making sprintf

- create a series of tests checking that the function is working correctly.

While implementing, I realized that the example described in the bug is just one of the possibilities (printing numbers) as by the Java specification, String.Format works with all types. I thus implemented the function to work with all the type that made sense in my opinion (we can discuss about this).

There is still one problem that I am not sure how to solve:
Test 04 and 07 and 08 depend on the user locale (it will be printed as 1.2.. or 1,2.. or with date the day or month depend on the locale). In order to solve this, we will need to show to String.Format the current locale but I could not find a way to do this from the XSDFuncOp class.

### CODE REVIEW

**Comment by afs** [comment]
Hi - looks good 

One suggestion : Java printf can take a variable number of arguments from zero upwards.

It would be nice if `afn:sprintf("%d != %d", 1, 2)` worked.  Could the arguments (args 1 on) go into a list, and javaSprintf proces them into a a Object[] for String.format?

**Comment by afs** [comment]
>  These cases for the moment are not supported. Maybe we could treat them all like strings.

Yes - `NodeFunction.str(NodeValue)` can be used to get the lexical form.

**Comment by ales004** [comment]
Yes I think that the idea about the list is very good. I will try to implement it like this.

I am still wondering what to do with the String.Format for Double, Float and Datetime that depend on the Locale. What do you think about this?

File: jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java [hunk 1]
```diff
@@ -352,6 +352,60 @@ public static NodeValue javaSubstring(NodeValue nvString, NodeValue nvStart, Nod
...
+                            return NodeValue.makeString(String.format(formatForOutput,decimalValue)) ;
+                        case OP_INTEGER:
+                            BigInteger integerValue = nvValue.getInteger();
+                            return NodeValue.makeString(String.format(formatForOutput,integerValue)) ;
+                        case OP_DOUBLE:
+                            Double doubValue = nvValue.getDouble();
+                            return NodeValue.makeString(String.format(formatForOutput,doubValue)) ;
+                        case OP_FLOAT:
+                            Float floatValue = nvValue.getFloat();
+                            return NodeValue.makeString(String.format(formatForOutput,floatValue)) ;
```
Comment (ales004) [inline jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java]: Depending on the Locale: 1.23 can be printed as 1.23 or 1,23

**Comment by afs** [comment]
For now, I think that "local locale" is the best single choice because if the user wants the XSD standard form "1.23" they can workaround it with a call `str()` and print a string. Not perfect but gives them a way.

XSD F&O v3.1 has "picture strings" : https://www.w3.org/TR/xpath-functions-3/#dt-picture-string : that looks like Java DecimalFormat, not printf style.

I think most people are familiar with printf and things like padding strings and alignment are covered. We can refine the capabilities after the machinery is in.

**Comment by ales004** [comment]
I think that commit bf14c982b8ae3eb8f040f55c28f8262efb5dad01 should do what you asked:

- now also tests are Locale dependent.
- sprintf accepts any number of arguments.
- unkwnown classes are translated to string using the function you pointed out.

What do you think?

**Comment by afs** [comment]
Looks good and I've merged it into master. Thanks!

(I've added t the documentation for the function library.)

**Pull request merged**